sel/src/Sel/Internal.hs

```diff
@@ -1,53 +1,103 @@
 {-# LANGUAGE BangPatterns #-}
-{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE ImportQualifiedPost #-}
 {-# LANGUAGE KindSignatures #-}
-{-# LANGUAGE MultiWayIf #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ViewPatterns #-}
 
 module Sel.Internal where
 
+import Control.Monad (when)
 import Control.Monad.IO.Class (MonadIO, liftIO)
-import qualified Data.Base16.Types as Base16
-import qualified Data.ByteString.Base16 as Base16
-import qualified Data.ByteString.Internal as BS
+import Control.Monad.Trans.Class (lift)
+import Data.Base16.Types qualified as Base16
+import Data.ByteString (StrictByteString)
+import Data.ByteString.Base16 qualified as Base16
+import Data.ByteString.Internal (memcmp)
+import Data.ByteString.Internal qualified as ByteString
+import Data.Coerce (coerce)
 import Data.Kind (Type)
-import Foreign (Ptr, castForeignPtr)
-import Foreign.C.Types (CInt (CInt), CSize (CSize))
-import Foreign.ForeignPtr (ForeignPtr, withForeignPtr)
-import LibSodium.Bindings.SecureMemory (sodiumFree, sodiumMalloc)
+import Foreign (ForeignPtr, Ptr)
+import Foreign qualified
+import Foreign.C (CSize, CUChar, throwErrno)
+import Foreign.C.Types (CChar)
+import LibSodium.Bindings.Comparison (sodiumMemcmp)
+import LibSodium.Bindings.SecureMemory (finalizerSodiumFree, sodiumFree, sodiumMalloc)
+import Sel.Internal.Scoped
+import Sel.Internal.Scoped.Foreign
 
--- | This calls to C's @memcmp@ function, used in lieu of
--- libsodium's @memcmp@ in cases when the return code is necessary.
-foreign import capi unsafe "string.h memcmp"
-  memcmp :: Ptr a -> Ptr b -> CSize -> IO CInt
+-- | Compare the contents of two byte arrays in constant time.
+--
+-- /See:/ [Constant-time test for equality](https://doc.libsodium.org/helpers#constant-time-test-for-equality)
+--
+-- @since 0.0.3.0
+foreignPtrEqConstantTime :: ForeignPtr CUChar -> ForeignPtr CUChar -> CSize -> IO Bool
+foreignPtrEqConstantTime p q size =
+  fmap (== 0) . use $
+    sodiumMemcmp <$> foreignPtr p <*> foreignPtr q <*> pure size
 
--- | Compare if the contents of two @ForeignPtr@s are equal.
-foreignPtrEq :: ForeignPtr a -> ForeignPtr a -> CSize -> IO Bool
-foreignPtrEq fptr1 fptr2 size =
-  withForeignPtr fptr1 $ \p ->
-    withForeignPtr fptr2 $ \q ->
-      do
-        result <- memcmp p q size
-        return $ 0 == result
+-- | Lexicographically compare the contents of two byte arrays.
+--
+-- ⚠️ Such comparisons are vulnerable to timing attacks, and should be
+-- avoided for secret data.
+--
+-- @since 0.0.1.0
+foreignPtrOrd :: ForeignPtr CUChar -> ForeignPtr CUChar -> CSize -> IO Ordering
```

Comment:
I'd like your opinion on this, but considering that we only manipulate data we manage, couldn't this function be of return type `Ordering` instead of `IO Ordering`?

Reply:
Certainly could be. In the existing code the practice seems to be to push the `unsafeDupablePerformIO` closer to the consumer, but I'm not at all opposed to just doing it here.

Reply:
Yeah let's not be too rigid, you can put the `unsafeDupablePerformIO` here. :)